### ISSUE TITLE
Validating self-referencing data structures

### ISSUE THREAD

**Issue by ehmicky** [ajv-validator/ajv#461]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**
5.0.4-beta.1

**Ajv options object**

```javascript
{}
```

**JSON Schema**

```json
{ "properties": { "self": { "$ref": "#" } } }
```

**Data**

```javascript
const data = {};
data.self = data;
```

**Your code**

```javascript
ajv.validate(schema, data);
```

**Validation result, data AFTER validation, error messages:**

```
RangeError: Maximum call stack size exceeded
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:29)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
    at validate (eval at localCompile (/node_modules/ajv/lib/compile/index.js:118:26), <anonymous>:3:389)
```

**What results did you expect?**
The validation should pass. Recursive structures should be handled. 

**Are you going to resolve the issue?**
No

**Comment by epoberezkin** [comment]
Validating ~~recursive~~ self-referencing data structures is out of scope of JSON-schema specification.

Whether to support it in Ajv depends on how common is the use case - could you explain what is it?

It would substantially complicate validation as it will have to keep track of all validated nodes and in case some node is encountered again and it has to be validated against the same schema it will be considered valid.

**Comment by epoberezkin** [comment]
Just to clarify, recursive $refs are fully supported, just not with self-referencing data structures. E.g. this data:

```json
{ "self": { "self": { "self": {} } } }
```

is valid according to this schema:

```json
{
  "type": "object",
  "properties": { "self": { "$ref": "#" } },
  "additionalProperties": false
}
```

**Comment by ehmicky** [comment]
I understand that supporting this would add a lot of complexity in the code, and probably degrade performance. I just thought I would bring it up, as this is being a problem for me right now, and you might be interested in knowing this issue arises for some users.

My use case is the following. I am working on an [open-source library](https://github.com/Coursio/api-engine) which generates an API server (both REST and GraphQL) using a single declarative file as input. This file describes each model, let's call it `models.schema.json`. Database CRUD, filtering, sorting, selecting, populating, validation, authorization, default attributes, etc. are automatically generated from parsing that file, without library users needing to write any other code.

I chose JSON schema as the format of `models.schema.json`, as it is expressive, well-designed, standard and has a rich ecosystem.

Database models often point to each other, or to themselves, creating circular references. I want to make sure `models.schema.json` is proper JSON schema before starting the server, so I validate it against [JSON schema meta-schema](http://json-schema.org/schema) (which has circular references too).

A simplified version of `models.schema.json` to represent a `User` would be:
```json
"required": ["id"],
"properties": {
   "id": {
      "type": "string"
   },
   "is_active": {
      "type": "boolean"
   },
   "best_friend": {
      "$ref": "#"
   }
}
```

A GraphQL query against this model, e.g. to find the best friend of a user's best friend, could look like this:
```graphql
findUser(id: 1) {
   best_friend {
      best_friend {
         id
      }
   }
}
```

I am dereferencing all JSON references prior to validation. I started by not doing so, but it turns out that most parts of my library are greatly simplified by having those JSON references dereferenced, except for `ajv` which crashes. An alternative for me is to not dereference me, but this would put back to other problems I encountered in my project by having to manually those JSON references every time some part of my system needs it.

**Comment by epoberezkin** [comment]
Interesting. You may be interested to read FAQ - the third part about $refs and why it is wrong/difficult to deference them in general case.

From my experience it's better to avoid self-referencing data structures that can't be serialised. You may save some time in the beginning but you are likely to find yourself in some tight corner further down the line. In general, there is bad support of self referencing data structures by most libraries... I'd stick with $refs.

**Comment by epoberezkin** [comment]
btw, supporting self-referencing schemas would be even more complex - atm it is a hard requirement that they are serialisable.

**Comment by ehmicky** [comment]
I understand. I did read that part of the FAQ first, but still wanted to bring it up just in case.

I think objects with circular references are still serializable, by first converting circular references to some kind of JSON-pointers-like string. This is e.g. what [this library](https://github.com/WebReflection/circular-json) I am using is doing. One could also dereference JSON references on parsing, and re-create them on serialization.

**Comment by ehmicky** [comment]
I ended up following up your advice, and avoiding circular references.

If this issue is too complex to implement, I think I'll close it. What do you think?

**Comment by maximelafarie** [comment]
Sorry to dig up this issue, but I'm in this case with the use of fastify and fastify/swagger, schemas being used for swagger and ajv at the same time. It's a shame the schema isn't valid for both (swagger supports self-referencing).

I don't know if this is something that's been permanently abandoned, but it would be so much more practical than having to redeclare similar structures.
